Make dl package thread-safe

alexbagirov opened the pull request:
This PR is intended to make `dl` calls thread-safe. Problem might occur if, for example, `dlopen()` and `dlerror()` were run in different goroutines and there was an error which happened after `dlopen()` but was ignored.

To fix this behaviour, each dl call (`dlopen()`, `dlsym()`, `dlclose()`) runs in the same thread with its `dlerror()` call. This is done using `runtime.LockOSThread()` wrapper around those calls.

elezar commented on pkg/dl/dl.go:
```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
 	}
 }
 
+func useLibDL(action func() error) error {
```
I don't think `useLibDL` is a good name for this function. What about `withOSLock()` or `withSync()`?

alexbagirov replied:
Renamed

elezar commented on pkg/dl/dl.go:
```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
+		return err
 	}
+
 	dl.handle = handle
```
Is setting (and clearing) the handle not also something that should happen under the lock?

alexbagirov replied:
Moved under lock

elezar commented on pkg/dl/dl.go:
```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
-	C.dlsym(dl.handle, sym)
-	err := C.dlerror()
-	if unsafe.Pointer(err) == C.NULL {
+	var pointer unsafe.Pointer
```
This pointer is local to the check below and could be initialised when calling `C.dlsym` below.

Also, we seem to have lost the call to `C.dlerror()` to clear previous errors. Is this no longer needed / expected?

alexbagirov replied:
I see no cases where we have to clear `dlerror` since we call it each time error actually happens (and now we know that we won't miss any errors since they will arrive in the same thread as `open`/`close`/`sym` calls). So imo this is no longer needed

alexbagirov replied:
Returned `dlError()` call.

elezar reviewed (changes requested):
Thanks for the PR. I have some minor comments / suggestions.

alexbagirov commented on pkg/dl/dl.go:
```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
+		return nil
+	}
+	return DLError{
+		err: C.GoString(lastErr),
```
Done

alexbagirov commented on pkg/dl/dl.go:
```diff
@@ -48,35 +49,80 @@ func New(name string, flags int) *DynamicLibrary {
+	return action()
+}
+
+type DLError struct {
```
Removed

alexbagirov commented:
Fixed all the comments and also added one more test for `dl` module. Added `t.Parallel()` to each test since now we're sure they can be run simultaneously.

elezar reviewed (dismissed):
Thanks @alexbagirov.

This looks good now.

elezar commented:
/blossom-ci

elezar commented:
/blossom-ci

anstockatnv commented:
/blossom-ci

elezar commented on pkg/dl/dl.go:
```diff
@@ -72,11 +102,16 @@ func (dl *DynamicLibrary) Lookup(symbol string) error {
-	if unsafe.Pointer(err) == C.NULL {
+	var pointer unsafe.Pointer
+	if err := withOSLock(func() error {
+		dlError()
```
Could we keep the comment here indicating the reason why we are calling `dlError()` here?
```suggestion
		// Call dlError() to clear out any previous errors.
		dlError()
```

alexbagirov replied:
Done
